src/Features/Core/Portable/Completion/Providers/TypeImportCompletionService.cs

```diff
@@ -0,0 +1,349 @@
+                    throw new ArgumentException(nameof(project));
+                }
+
+                var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+                var checksum = await SymbolTreeInfo.GetSourceSymbolsChecksumAsync(project, cancellationToken).ConfigureAwait(false);
+
+                GetAccessibleTopLevelTypesWorker(
+                    project.Id,
+                    compilation.Assembly.GlobalNamespace,
+                    checksum,
+                    isInternalsVisible: true,
+                    handleAccessibleItem,
+                    _projectItemsCache,
+                    cancellationToken);
+            }
+
+            public async Task GetAccessibleTopLevelTypesFromCompilationReferenceAsync(
+                Solution solution,
+                Compilation compilation,
+                CompilationReference compilationReference,
+                Action<CompletionItem> handleAccessibleItem,
+                CancellationToken cancellationToken)
+            {
+                if (!(compilation.GetAssemblyOrModuleSymbol(compilationReference) is IAssemblySymbol assemblySymbol))
+                {
+                    return;
+                }
+
+                var isInternalsVisible = compilation.Assembly.IsSameAssemblyOrHasFriendAccessTo(assemblySymbol);
+                var assemblyProject = solution.GetProject(assemblySymbol, cancellationToken);
+                var checksum = await SymbolTreeInfo.GetSourceSymbolsChecksumAsync(assemblyProject, cancellationToken).ConfigureAwait(false);
+
+                GetAccessibleTopLevelTypesWorker(
+                    assemblyProject.Id,
+                    assemblySymbol.GlobalNamespace,
+                    checksum,
+                    isInternalsVisible,
+                    handleAccessibleItem,
+                    _projectItemsCache,
+                    cancellationToken);
+            }
+
+            public void GetAccessibleTopLevelTypesFromPEReference(
+                Solution solution,
+                Compilation compilation,
+                PortableExecutableReference peReference,
+                Action<CompletionItem> handleAccessibleItem,
+                CancellationToken cancellationToken)
+            {
+                if (!(compilation.GetAssemblyOrModuleSymbol(peReference) is IAssemblySymbol assemblySymbol))
+                {
+                    return;
+                }
+
+                var key = GetReferenceKey(peReference);
+                var isInternalsVisible = compilation.Assembly.IsSameAssemblyOrHasFriendAccessTo(assemblySymbol);
+                var rootNamespaceSymbol = assemblySymbol.GlobalNamespace;
+
+                if (key == null)
+                {
+                    // Can't cache items for reference with null key, so just create them and return. 
+                    var items = GetCompletionItemsForTopLevelTypeDeclarations(rootNamespaceSymbol, cancellationToken);
```

Comment:
i'm worried about the perf implications of this in practice.  i.e. we hit some reference that doesn't have a key, and we keep on making items over and over and over again.  i would bail in this case instead of falling back to code that will have a perf profile we don't udnerstand.

Reply:
Makes sense.

Reply:
I have changed this as you suggested. But would also hear @dpoeschl and @jasonmalinowski 's thought on this